ddl/ddl_api.go

```diff
@@ -1161,6 +1163,67 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 	return errors.Trace(err)
 }
 
+func (d *ddl) AddPartitions(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTableSpec) error {
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ti.Schema)
+	if !ok {
+		return errors.Trace(infoschema.ErrDatabaseNotExists.GenByArgs(schema))
+	}
+	t, err := is.TableByName(ti.Schema, ti.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableExists.GenByArgs(ti.Schema, ti.Name))
+	}
+	if len(spec.Definitions) == 0 {
+		return errors.Trace(infoschema.ErrPartitionsMustBeDefined)
+	}
+	// TODO: Check that the add value must be greater than all previous values.
+
+	par := &model.PartitionInfo{}
+	if t.Meta().Partition != nil {
+		pi := &model.PartitionInfo{
+			Type: t.Meta().Partition.Type,
+			Expr: t.Meta().Partition.Expr,
+		}
+		for _, def := range spec.Definitions {
+			pid, err1 := d.genGlobalID()
+			if err1 != nil {
+				return errors.Trace(err1)
+			}
+			piDef := model.PartitionDefinition{
+				Name:    def.Name,
+				ID:      pid,
+				Comment: def.Comment,
+			}
+			for _, expr := range def.LessThan {
+				buf := new(bytes.Buffer)
+				expr.Format(buf)
+				piDef.LessThan = append(piDef.LessThan, buf.String())
+			}
+			pi.Definitions = append(pi.Definitions, piDef)
+		}
+		oldDef := t.Meta().Partition.Definitions
```

Comment:
We should make a copy of the slice to avoid modify the slice in the meta.
```
par.Definitions := make([]model.PartitionDefinition, 0, len(t.Meat().Partition.Definitions)+len(spec.Definitions))
```

Then append old and new partitions.